force-app/main/default/classes/SOQL.cls

```diff
@@ -270,92 +356,182 @@ public with sharing class SOQL {
+    for (Id targetOpportunityId : cloneOppIds) {
+        // For each source line item, create a clone for the target opportunity
+        for (OpportunityLineItem sourceItem : sourceLineItems) {
+            OpportunityLineItem clonedItem = sourceItem.clone(false, true); // shallow clone, preserve read-only fields
+            clonedItem.OpportunityId = targetOpportunityId; // Set the target opportunity Id
+            clonedLineItems.add(clonedItem); // Add to the list for bulk insertion
+        }
+    }
 
-        // Insert the new line items
+    // Insert the cloned line items in bulk
+    if (!clonedLineItems.isEmpty()) {
+        insert clonedLineItems;
     }
+} 
     
     /**
      * Question 24
      * Retrieves the Opportunity Name, Account Name, and Owner Name for Opportunities that have a Stage of 'Closed Won'.
      * Example: getOpportunityDetailsForClosedWon()
      * @return A list of filtered Opportunity
      */
-    public static List<Opportunity> getOpportunityDetailsForClosedWon() {        
-        return null; // Replace null with the variable you used to store the result
+    public static List<Opportunity> getOpportunityDetailsForClosedWon() {  
+        List<Opportunity> closedWonOpportunities = [ 
+            SELECT Id, Name, Account.Name, Owner.Name
+            FROM Opportunity
+            WHERE StageName = 'Closed Won'
+        ];     
+        return closedWonOpportunities; // Replace null with the variable you used to store the result
     }
 
     /**
      * Question 25
      * Sets the 'Primary_Contact__c' field for all Closed Won Opportunities to first Contact on the Account.
      * Example: setPrimaryContact() should set the 'Primary_Contact__c' field for every Opportunity in the org
      */
+     
     public static void setPrimaryContact() {
         //Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
-
+        List<Opportunity> closedWonOpportunities = getOpportunityDetailsForClosedWon();
         // Create a set of Account Ids from the Opportunities
+        Set<Id> accountIds = new Set<Id>();
+        for (Opportunity opp : closedWonOpportunities) {
+            if (opp.AccountId != null) {
+                accountIds.add (opp.AccountId );
+            }
+        }
+        
+        List<Contact> contacts = new List<Contact>();
 
         // Query the Contacts
+        if (!accountIds.isEmpty()) {
+            contacts = [
+                SELECT Id, AccountId
+                FROM Contact
+                WHERE AccountId IN :accountIds
+            ];
+        }
 
         // Loop through each Contact and add the Account Id and Contact to the map
+        Map<Id,Contact> accountToContactMap = new Map<Id,Contact> ();
+        for (Contact contact : contacts) {
+            if (!accountToContactMap.containsKey(contact.AccountId)){
+                accountToContactMap.put(contact.AccountId, contact);
+            }
+        }
```

Comment:
What happens if an Account has more than one Contact? When a new Contact is added to the map, what happens to the previously added Contact (if they share the same Account)?

(Honestly, this method's instructions are a bit confusing. What does it mean by the first Contact on the Account?)